pkg/apis/core/v1beta1/helper/helper.go

```diff
@@ -782,3 +699,120 @@ func FindPrimaryDNSProvider(providers []gardencorev1beta1.DNSProvider) *gardenco
+	v, err := semver.NewVersion(currentVersion)
+	if err != nil {
+		return false, "", err
+	}
+	currentSemVerVersion := *v
+
+	predicate := func(v semver.Version) (bool, error) {
+		if v.Equal(&currentSemVerVersion) {
+			return true, nil
+		}
+		isWithinRange, err := versionutils.CompareVersions(v.String(), "~", currentSemVerVersion.String())
+		if err != nil {
+			return true, err
+		}
+		return !isWithinRange, nil
+	}
+
+	latestVersion, err := GetLatestQualifyingVersion(cloudProfile.Spec.Kubernetes.Versions, predicate)
+	if err != nil {
+		return false, "", nil
+	}
+
+	return true, latestVersion.Version, nil
+}
+
+// GetKubernetesVersionForMinorUpdate finds a Kubernetes version in the <cloudProfile> that qualifies for a Kubernetes minor level update given a <currentVersion>.
+// A qualifying version is a non-preview version having the minor version increased by exactly one version.
+// If a version can be found, returns true and the latest patch version of the next minor version.
+// In case it does not find a version, it returns false.
+func GetKubernetesVersionForMinorUpdate(cloudProfile *gardencorev1beta1.CloudProfile, currentVersion string) (bool, string, error) {
+	v, err := semver.NewVersion(currentVersion)
+	if err != nil {
+		return false, "", err
+	}
+	currentSemVerVersion := *v
+
+	predicate := func(v semver.Version) (bool, error) {
+		if v.Equal(&currentSemVerVersion) {
+			return true, nil
+		}
+		isWithinRange, err := versionutils.CompareVersions(v.String(), "^", currentSemVerVersion.String())
+		if err != nil {
+			return true, err
+		}
+
+		if !isWithinRange {
+			return true, nil
+		}
+
+		hasIncorrectMinor := currentSemVerVersion.Minor()+1 != v.Minor()
+		return hasIncorrectMinor, nil
+	}
+
+	latestVersion, err := GetLatestQualifyingVersion(cloudProfile.Spec.Kubernetes.Versions, predicate)
+	if err != nil {
+		return false, "", nil
+	}
+
+	return true, latestVersion.Version, nil
+}
+
+// FilterQualifyingVersions removes versions from a set of expirable versions that do not qualify for updates during maintenance time
+// Removes preview versions, expired versions and versions based on a predicate
+// filters a version if the predicate returns true
+// returns a map[version]ExpirableVersion and a set of versions containing the filtered versions
+func FilterQualifyingVersions(versions []gardencorev1beta1.ExpirableVersion, predicate func(a semver.Version) (bool, error)) (map[string]gardencorev1beta1.ExpirableVersion, sets.String, error) {
+	filteredVersions := make(map[string]gardencorev1beta1.ExpirableVersion, len(versions))
+	filtered := sets.String{}
+	for _, v := range versions {
+		if v.Classification != nil && *v.Classification == gardencorev1beta1.ClassificationPreview {
+			continue
+		}
+
+		if v.ExpirationDate != nil && (time.Now().UTC().After(v.ExpirationDate.UTC()) || time.Now().UTC().Equal(v.ExpirationDate.UTC())) {
```

Comment:
I am not sure, but I think the behaviour as described [here](https://github.com/gardener/gardener/blob/aaca736472c700f9cd9f8b4dbec06b9d2f4f90c0/docs/operations/versioning.md#forceful-migration-of-expired-versions) will not work:
> If that's expired as well, the update process repeats until a non-expired Kubernetes version is reached.
Depending on the circumstances described above, it can happen that the cluster receives multiple consecutive minor Kubernetes version updates! 

How will you do two consecutive minor upgrades (with the first upgrade being to an expired version) if you filter out all expired versions?